src/runtime-rs/crates/runtimes/common/src/runtime_handler.rs

```diff
@@ -41,3 +42,27 @@ pub trait RuntimeHandler: Send + Sync {
 
     fn cleanup(&self, id: &str) -> Result<()>;
 }
+
+impl RuntimeInstance {
+    // NOTE: if static resource management is configured, a warning is logged
+    // hotplug vcpu/memory, and the cpu will not be updated since the sandbox
+    // resource should be static
+    //
+    // The updated resource is calculated from:
+    //   - vcpu: the sum of each ctr, plus default vcpu
+    //   - memory: the sum of each ctr, plus default memory, and setup swap
+    pub async fn update_sandbox_resource(&self) -> Result<()> {
```

Comment:
I think `update_sandbox_resource` only works with `virt-containers`. If the runtime instance is Linux or wasm, I think it can't work. My suggestion is to implement it in `container_manager`, because you can also fetch `resource_manager`  there, then you don't need to take it from` virt_sandbox`

Reply:
> I think the implementation of `update_sandbox_resource` only works with `virt-containers`. If the runtime instance is Linux or wasm, I think it can't work. My suggestion is to implement it in `container_manager`, because you can also fetch `resource_manager` there, then you don't need to take it from` virt_sandbox`

Reply:
But my point is this operation only works for virt-containers, if the runtime instance is Linux or wasm, this won't work

Reply:
> But my point is this operation only works for virt-containers, if the runtime instance is Linux or wasm, this won't work

This works with other type of container, since a sandbox often manages more than one containers.

Reply:
`let nr_vcpus = self.container_manager.vcpus_required().await?;` I think only virt-container would have vcpu

Reply:
> `let nr_vcpus = self.container_manager.vcpus_required().await?;` I think only virt-container would have vcpu

This `vcpu` is actually the rounded `cpu_quota / cpu_period` value, this semantic applys to all kinds of container. Do you think changing this to `cpu_required()` will be fine?

Reply:
yes, I agree

Reply:
@Tim-0731-Hzt I agree with @Ji-Xinyou that this is the resource of a `sandbox` rather than a container, sandbox resource is the summary of the sandbox, it works for Virt-container/Linux-container/Wasm-container. For Virt-container it is the host resource such as CPU and memory cgroup. For Linux-container it is the Pod Level resource such as CPU, and memory cgroup. 
All the differences between Virt-container/Linux-container/Wasm-container are implemented in the Sandbox.

For the design, I have some suggestions @Ji-Xinyou 
1. The `update_sandbox_resource()` should be implemented in the sandbox, the sandbox can get the total CPU from the cgroup resource, which saves all the Linux resources for all containers.
2. The design of `CpuMemResource` is a little thin, and it is recommended to combine it with `CgroupsResource` to look appropriate

Reply:
@quanweiZhou @Ji-Xinyou I think I didn't express my opinion clearly. My concern is if you call `instance.update_sandbox_resource()`, if the runtime instance is virt-container, the current design works fine. If the runtime instance is wasm or Linux, `update_sandbox_resource()` may not work the same way as virt-container. wasm and Linux might have different implementations for `update_sandbox_resource()`

Reply:
So my suggestion is to implement `update_sandbox_resource` for different runtime instance

Reply:
Please correct me if I am wrong. IMO, we have an abstraction, `ResourceManager`, which is independent from `Sandbox` and `ContainerManager`. I think we can implement the real `update_sandbox_resource()` in `ResourceManager`. Calling this function from `Sandbox` or `ContainerManager` are all ok for me. (If we calling the function from ContainerManager, we can use a sandbox_id parameter.)